rest/main.go

```diff
@@ -146,3 +173,158 @@ func serverMainPersistentConfig(fs *flag.FlagSet, flagStartupConfig *StartupConf
+// Returns the new startup config, a bool of whether to fallback to legacy config and an error
+func automaticConfigUpgrade(configPath string) (*StartupConfig, bool, error) {
+	legacyServerConfig, err := LoadServerConfig(configPath)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if legacyServerConfig.DisablePersistentConfig != nil && *legacyServerConfig.DisablePersistentConfig {
+		return nil, true, nil
+	}
+
+	startupConfig, dbConfigs, err := legacyServerConfig.ToStartupConfig()
+	if err != nil {
+		return nil, false, err
+	}
+
+	dbConfigs, err = sanitizeDbConfigs(dbConfigs)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Attempt to establish connection to server
+	cluster, err := establishCouchbaseClusterConnection(startupConfig)
+	if err != nil {
+		return nil, false, err
+	}
+
+	defer func() {
+		_ = cluster.Close()
+	}()
+
+	// Write database configs to CBS with groupID "default"
+	for _, dbConfig := range dbConfigs {
+		_, err = cluster.PutConfig(*dbConfig.Bucket, persistentConfigDefaultGroupID, base.Uint64Ptr(0), dbConfig)
+		if err != nil {
+			// TODO: if exists skip, else error
+			return nil, false, err
+		}
+	}
+
+	// Attempt to backup current config
+	err = backupCurrentConfigFile(configPath)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Overwrite old config with new migrated startup config
+	jsonStartupConfig, err := json.Marshal(startupConfig)
+	if err != nil {
+		return nil, false, err
+	}
+
+	err = ioutil.WriteFile(configPath, jsonStartupConfig, 0644)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return startupConfig, false, nil
+}
+
+// validate / sanitize db configs
+// - remove servers
+// - remove users
+// - ensure servers are the same
+func sanitizeDbConfigs(configMap DbConfigMap) (DbConfigMap, error) {
+	var databaseServerAddress string
+
+	for dbName, dbConfig := range configMap {
+		if databaseServerAddress == "" {
+			databaseServerAddress = *dbConfig.Server
+		}
+
+		if *dbConfig.Server != databaseServerAddress {
+			return nil, fmt.Errorf("automatic upgrade to persistent config requires matching server addresses in " +
+				"2.x config")
+		}
+
+		if dbConfig.Bucket == nil || *dbConfig.Bucket == "" {
+			dbNameCopy := dbName
+			dbConfig.Bucket = &dbNameCopy
+		}
+
+		dbConfig.Name = dbName
+
+		dbConfig.Server = nil
+		dbConfig.Users = nil
+		dbConfig.Roles = nil
+
+		// Make sure any updates are written back to the config
+		configMap[dbName] = dbConfig
+	}
+	return configMap, nil
+}
+
+// backupCurrentConfigFile takes the original config path and copies this to a file with -bk appended with a timestamp
+func backupCurrentConfigFile(sourcePath string) error {
+	source, err := os.Open(sourcePath)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = source.Close()
+	}()
+
+	// Grab file extension
+	fileExtension := filepath.Ext(sourcePath)
+
+	// Remove file extension so we can easily modify filename
+	fileNameWithoutExtension := strings.ReplaceAll(sourcePath, fileExtension, "")
```

Comment:
This seems safer...

```suggestion
	// Remove file extension so we can easily modify filename
	fileNameWithoutExtension := strings.TrimSuffix(sourcePath, fileExtension)
```

Reply:
I was looking for a "ReplaceLast" or something....